Lib/logging/__init__.py

```diff
@@ -1555,36 +1555,41 @@ def findCaller(self, stack_info=False, stacklevel=1):
         Find the stack frame of the caller so that we can note the source
         file name, line number and function name.
         """
-        f = currentframe()
+        f = _currentframe()
         #On some versions of IronPython, currentframe() returns None if
         #IronPython isn't run with -X:Frames.
-        if f is not None:
-            f = f.f_back
-        orig_f = f
-        while f and stacklevel > 1:
-            f = f.f_back
-            stacklevel -= 1
-        if not f:
-            f = orig_f
-        rv = "(unknown file)", 0, "(unknown function)", None
-        while hasattr(f, "f_code"):
+        if f is None:
+            return "(unknown file)", 0, "(unknown function)", None
+        while stacklevel > 0:
+            next_f = f.f_back
+            if next_f is None:
+                ##TODO: We've got options here
+                ## If we want to use the last (deepest) frame:
+                break
+                ## If we want to mimic the warnings module:
+                #return ("sys", 1, "(unknown function)", None)
+                ## If we want to be pedantic:
+                #raise ValueError("call stack is not deep enough")
+            f = next_f
             co = f.f_code
             filename = os.path.normcase(co.co_filename)
-            if filename == _srcfile:
-                f = f.f_back
+            # The following condition is partially copied from
+            # warnings._is_internal_frame
+            if filename == _srcfile or (
+                "importlib" in filename and "_bootstrap" in filename):
+            ):
                 continue
-            sinfo = None
-            if stack_info:
-                sio = io.StringIO()
-                sio.write('Stack (most recent call last):\n')
-                traceback.print_stack(f, file=sio)
-                sinfo = sio.getvalue()
-                if sinfo[-1] == '\n':
-                    sinfo = sinfo[:-1]
-                sio.close()
-            rv = (co.co_filename, f.f_lineno, co.co_name, sinfo)
-            break
-        return rv
+            stacklevel -= 1
+        sinfo = None
+        if stack_info:
+            sio = io.StringIO()
```

Comment:
Is there a context manager for StringIO()? It would be better than calling close().